### DISCUSSION TITLE
aggregating tuples

### DISCUSSION THREAD

**Discussion by ktlina** [nextflow-io/nextflow#5425] [General]
I have a workflow that executes different processes depending on the users preferences. When all desired processes are complete, a last process will take all the outputs and create a summary of the most relevant results. To do this, I need to take the output tuples generated by all processes and aggregate them into a single tuple. So for instance:

1. Process 1 output: ["sample1", /home/output/process1/sample1]
2. Process 2 output: ["sample1", /home/output/process2/sample1]
3. Process 3 output: ["sample1", /home/output/process3/sample1]

When using mix() and groupTuple() I am able to group the first two tuples and the result looks something like:
["sample1", [/home/output/process1/sample1, /home/output/process2/sample1]]. However, when I try to mix and grouTuple for the third process, I get this error:
**'Not a valid path value type: nextflow.util.ArrayBag'**
Mind you I'm trying to mix the resulting tuple (Tuple1/2) of mixing Tuple1 and Tuple2 with Tuple3. 

I understand this is because I am trying to use an Array/List as the mix() method input, and this is not supported. So my question is, how can I group the process outputs in a tuple containing the sampleID and a list of paths? I need this in order to make the summary process as "relaxed" as possible, so that, even though the user decides not to run all processes, it will still work. This way I do not need to create an input for the summary process that looks like: tuple val(sampleID), path(process1), path(process2), path(process3) but instead, **tuple val(sampleID), path(all_results)**. 

I hope I have explained myself clearly. 
Thnk you in advance

**Comment by bentsherman** [comment]
Can you share a minimal code example?

**Reply by ktlina** [reply]
Sorry, I just saw this. I kind of solved the problem by using "join" method and then mapping the tuples in the format I want them to be... so joining the tuple outputs of different processes (all_files = all_files.join(runMegaHitOutput, by: 0)) and then mapping them         (final_tuple = all_files.map {tuple -> return [tuple[0], tuple[1..-1]]}) into a final tuple.

**Reply by bentsherman** [reply]
Glad you found a solution. `join` is probably what I would recommend for grouping related outputs from different processes, whereas `groupTuple` would be more for grouping related outputs from the same process. We are working to stamp out some of these quirks with the different operators so that it isn't such a hassle

**Reply by ktlina** [reply]
yeas, i feel like i have been having a lot ouf trouble when dealing with the combination of outputs from different processes, especially when working with multiple samples. thank you for you help, anyways!

**Reply by jiading** [reply]
Hi @ktlina , how did you solve the problem?

**Reply by ktlina** [reply]
I work with an all_files tuple (sample, files) to which I join the output of each process: `all_files = all_files.join(processOutput)` at the end of said processes (important the tuple name must always be the same so it is updated with each process). This way each process output is appended to the `all_files` tuple independent of how many processes the user chooses to run. At the end when I'm ready to pass the tuple to my final summarization process, I map the files like this: `final_tuple = all_files.map {tuple ->return [tuple[0], tuple[1..-1]]}` so that the second element of my tuple, contains as many process outputs as processes the user decided to run. 
Hope this helps!
